packages/core/src/types/config/index.ts

```diff
@@ -20,76 +20,105 @@ import type {
-      kind: 's3'
-      /** Sets signing of the asset - for use when you want private assets */
-      signed?: { expiry: number }
-      generateUrl?: (path: string) => string
-      /** Sets whether the assets should be preserved locally on removal from keystone's database */
-      preserve?: boolean
-      pathPrefix?: string
-      /** Your s3 instance's bucket name */
-      bucketName: string
-      /** Your s3 instance's region */
-      region: string
-      /** An access Key ID with write access to your S3 instance */
-      accessKeyId?: string
-      /** The secret access key that gives permissions to your access Key Id */
-      secretAccessKey?: string
-      /** An endpoint to use - to be provided if you are not using AWS as your endpoint */
-      endpoint?: string
-      /** If true, will force the 'old' S3 path style of putting bucket name at the start of the pathname of the URL  */
-      forcePathStyle?: boolean
-      /** A string that sets permissions for the uploaded assets. Default is 'private'.
-       *
-       * Amazon S3 supports a set of predefined grants, known as canned ACLs.
-       * See https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html#canned-acl
-       * for more details.
-       */
-      acl?:
-        | 'private'
-        | 'public-read'
-        | 'public-read-write'
-        | 'aws-exec-read'
-        | 'authenticated-read'
-        | 'bucket-owner-read'
-        | 'bucket-owner-full-control'
-    }
+    /** The kind of storage being configured */
+    kind: 's3'
+    /** Sets signing of the asset - for use when you want private assets */
+    signed?: { expiry: number }
+    generateUrl?: (path: string) => string
+    /** Sets whether the assets should be preserved locally on removal from keystone's database */
+    preserve?: boolean
+    pathPrefix?: string
+    /** Your s3 instance's bucket name */
+    bucketName: string
+    /** Your s3 instance's region */
+    region: string
+    /** An access Key ID with write access to your S3 instance */
+    accessKeyId?: string
+    /** The secret access key that gives permissions to your access Key Id */
+    secretAccessKey?: string
+    /** An endpoint to use - to be provided if you are not using AWS as your endpoint */
+    endpoint?: string
+    /** If true, will force the 'old' S3 path style of putting bucket name at the start of the pathname of the URL  */
+    forcePathStyle?: boolean,
+    /** The configuration for keystone's hosting of the assets - if set to null, keystone will not host the assets */
+    serverRoute: {
+      /** The partial path that the assets will be hosted at by keystone, eg `/images` or `/our-cool-files` */
+      path: string,
+    } | null,
+    /** A function that is checked before serving the file or image to check for permissions. 
+     * This function will only be respected if the serverRoute is set
+    */
+    isAccessAllowed?: (options: StorageAccessAllowedOptions) => boolean,
```

Comment:
If this only works if `serverRoute` is set, we should probably put this in `serverRoute`